src/hb/shape_wasm.rs

```diff
@@ -0,0 +1,421 @@
+        // From HarfBuzz docs:
+        // (In the following functions, a font is a specific instantiation of a face at a
+        // particular scale factor and variation position.)
+        //
+        // I am unsure how to represent that in rustybuzz.
+        // er .. do what here?
+        0
+    };
+    linker
+        .func_wrap(module_name, "font_get_face", font_get_face)
+        .ok()?;
+
+    // fn font_get_glyph(font: u32, unicode: u32, uvs: u32) -> u32;
+    // Returns the nominal glyph ID for the given codepoint, using the cmap table of the font to map Unicode codepoint (and variation selector) to glyph ID.
+    let font_get_glyph =
+        |mut caller: Caller<'_, ShapingData>, _: u32, codepoint: u32, uvs: u32| -> u32 {
+            let Some(codepoint) = char::from_u32(codepoint) else {
+                return 0;
+            };
+            let Some(uvs) = char::from_u32(uvs) else {
+                return 0;
+            };
+            caller
+                .data()
+                .font
+                .glyph_variation_index(codepoint, uvs)
+                .unwrap_or_default()
+                .0 as u32
+        };
+    linker
+        .func_wrap(module_name, "font_get_glyph", font_get_glyph)
+        .ok()?;
+
+    // fn font_get_scale(font: u32, x_scale: *mut i32, y_scale: *mut i32);
+    // Returns the scale of the current font.
+    // Just return the upem as rustybuzz has no scale.
+    let font_get_scale =
+        |mut caller: Caller<'_, ShapingData>, _: u32, x_scale: u32, y_scale: u32| {
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+            let upem = caller.data().font.units_per_em();
+
+            memory.write(
+                &mut caller.as_context_mut(),
+                x_scale as usize,
+                &upem.to_le_bytes(), // le or be?
+            );
+            memory.write(
+                &mut caller.as_context_mut(),
+                y_scale as usize,
+                &upem.to_le_bytes(),
+            );
+        };
+    linker
+        .func_wrap(module_name, "font_get_scale", font_get_scale)
+        .ok()?;
+
+    // fn font_get_glyph_extents(font: u32, glyph: u32, extents: *mut CGlyphExtents) -> bool;
+    // Returns the glyph's extents for the given glyph ID at current scale and variation settings.
+    let font_get_glyph_extents =
+        |mut caller: Caller<'_, ShapingData>, _: u32, glyph: u32, extents: u32| -> u32 {
+            let memory = caller.get_export("memory").unwrap().into_memory().unwrap();
+
+            let mut glyph_extents = hb_glyph_extents_t::default();
+            let ret = caller
+                .data()
+                .font
+                .glyph_extents(GlyphId(glyph as u16), &mut glyph_extents);
+            if ret {
+                // WASM is little endian. MacOS is little endian, I think.
+                // Maybe should use bytemuck as a dependency instead.
```

Comment:
We already use `bytemuck`

Reply:
Hah right. I need to read up on the API.